engine/execution/ingestion/engine.go

```diff
@@ -930,344 +992,422 @@ func (e *Engine) generateExecutionReceipt(
-		if errors.Is(getStateCommitmentErr, storage.ErrNotFound) {
-			return nil
-		}
+	lastSealed, err := e.state.Sealed().Head()
+	if err != nil {
+		return fmt.Errorf("could not get last sealed: %w", err)
+	}
 
-		// sanity check that if the state for the parent block already exists,
-		// then the delta's start state must be equal to that.
-		if !bytes.Equal(stateCommitment, executionStateDelta.StartState) {
-			return fmt.Errorf("internal inconsistency with delta for block (%v) - state commitment for parent retrieved from DB (%x) different from start state in delta! (%x)",
-				executionStateDelta.ParentID(),
-				stateCommitment,
-				executionStateDelta.StartState)
-		}
+	sealedHeight := lastSealed.Height
 
-		// if the parent state exists for the orphan delta, we could apply this delta.
-		e.syncWg.Add(1)
-		go e.saveDelta(ctx, executionStateDelta)
+	log := e.log.With().
+		Hex("sender", originID[:]).
+		Uint64("sealed", sealedHeight).
+		Uint64("from", req.FromHeight).
+		Uint64("to", req.ToHeight).
+		Logger()
 
+	// ignore requests for unsealed height
+	if req.FromHeight > sealedHeight {
+		log.Info().Msg("receives state sync requests for unsealed height, ignore")
 		return nil
-	})
+	}
+
+	// fromHeight, toHeight must be sealed height
+	fromHeight, toHeight := req.FromHeight, req.ToHeight
+	if toHeight > sealedHeight {
+		toHeight = sealedHeight
+	}
+
+	// for each height starting from fromHeight to toHeight,
+	// query the statecommitment, and if exists, send the
+	// state delta
+	// TOOD: add context
+	ctx := e.unit.Ctx()
+	err = e.deltaRange(ctx, fromHeight, toHeight,
+		func(delta *messages.ExecutionStateDelta) {
+			err := e.syncConduit.Submit(delta, originID)
+			if err != nil {
+				e.log.Error().Err(err).Msg("could not submit block delta")
+			}
+		})
+
+	if err != nil {
+		return fmt.Errorf("could not send deltas: %w", err)
+	}
+
+	log.Info().Msg("responded state deltas for a height range")
+
+	return nil
 }
 
-func (e *Engine) saveDelta(ctx context.Context, executionStateDelta *messages.ExecutionStateDelta) {
+// deltaRange querys it's local execution state to find deltas for a height
+// range between the fromHeight to the toHeight. If delta is found, then
+// pass it to the onDelta callback.
+func (e *Engine) deltaRange(ctx context.Context, fromHeight uint64, toHeight uint64,
+	onDelta func(*messages.ExecutionStateDelta)) error {
+
+	for height := fromHeight; height <= toHeight; height++ {
+		header, err := e.state.AtHeight(height).Head()
+		if err != nil {
+			return fmt.Errorf("could not query block header at height: %v", height)
+		}
 
-	defer e.syncWg.Done()
+		blockID := header.ID()
+		_, err = e.execState.StateCommitmentByBlockID(ctx, blockID)
 
+		if err == nil {
+			// this block has been executed, we will send the delta
+			delta, err := e.execState.RetrieveStateDelta(ctx, blockID)
+			if err != nil {
+				return fmt.Errorf("could not retrieve state delta for block %v, %w", blockID, err)
+			}
+
+			onDelta(delta)
```

Comment:
Should we ignore error here?

Reply:
The error is already handled [here](https://github.com/onflow/flow-go/pull/31/files#diff-9d11a07967b04bd9f51ebd6b01a170d5R1200-R1202)